### PR TITLE
Cache option for searchers

### LINKED ISSUE #42: Adding caching support
As mentioned in the README, lilconfig doesn't currently support caching. It would be possible to greatly improve Prettier CLI performance if lilconfig were to implement some caching capabilities. Based on the analysis I have done, adding caching support could potentially improve Prettier CLI performance by ~15% on a large-scale project. More details can be found here: https://github.com/prettier/prettier/issues/15653.

@antonk52 Are you open to adding caching capabilities to lilconfig in a similar fashion as cosmiconfig? I wanted to check with you first before I spend time working on this.

### PR DESCRIPTION
**Pull request by antonk52**
This PR introduces a new option `cache`(it should have been there from the start). The API and implementation is inspired by cosmiconfig. Similarities between lilconfig and cosmiconfig implementation:

- Both use maps for storing caching results
- The same public API to clear searchCache/loadCache/both
- `cache` option is enabled by default
- sync and async load methods use filepath argument as cache key ([cosmiconfig](https://github.com/cosmiconfig/cosmiconfig/blob/42ca3fab6d6ae593a895c0fe4e2a5f6b297e6361/src/Explorer.ts#L22C44-L22C52))
- sync and async search methods use path to a directory where lookup for search places will take place as cache key ([cosmiconfig](https://github.com/cosmiconfig/cosmiconfig/blob/42ca3fab6d6ae593a895c0fe4e2a5f6b297e6361/src/Explorer.ts#L22C44-L22C52) from is reassigned and called recursively in a closure)

I have added test cases for cache option enabled and disabled. Jest reports test coverage to be 100%. I am happy to add more tests if anyone can spot any missed cases.

Once approved I will merge this PR and release this as a new major version since this option is enabled by default. I consider this change in behaviour a breaking change. A few more changes will be merged prior to releasing a new version like dropping support for legacy node.js versions.

TODO
- [x] async implementation
- [x] sync implementation
- [x] test enabled option
- [x] test disabled option
- [x] add to readme
- [x] compare to cosmiconfig implementation
- [x] add searchers shape comparison in tests

fixes #42

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -204,7 +228,22 @@ export function lilconfig(
...
+            const visited: Set<string> = new Set();
+            for (const {
+                searchPlace,
+                filepath,
+                loaderKey,
+                searchPath,
+            } of searchItems) {
+                if (cache) {
+                    if (searchCache.has(searchPath)) {
+                        return searchCache.get(searchPath) as
```
Comment (pmdartus) [inline src/index.ts:240]: On a cache hit, shouldn't it break the for loop instead of directly returning the result from the cache? 

As far as I understand the code, if there is a cache hit while walking the tree upward, the search cache will not be populated with the previously visited paths.

For example, if there is a single configuration file present at the root of the project.

```js
resolver.search('./a');    // 1
resolver.search('./a/b');  // 2
resolver.search('./a/b');  // 3
```

- After 1, the `searchCache` will have the following keys: `./` and `./a`
- After 2, the `searchCache` should have the following keys: `./`, `./a`, and `./a/b`. But with the current implementation it will only have: `./` and `./a`
- As a result in 3, when invoking `search` on the same path lilconfig will still have to lookup on the file system for config present at `./a/b`

File: src/index.ts [hunk 2]
```diff
@@ -204,7 +228,22 @@ export function lilconfig(
...
-            for (const {searchPlace, filepath, loaderKey} of searchItems) {
+            const visited: Set<string> = new Set();
+            for (const {
+                searchPlace,
+                filepath,
+                loaderKey,
+                searchPath,
+            } of searchItems) {
+                if (cache) {
+                    if (searchCache.has(searchPath)) {
```
Comment (pmdartus) [inline src/index.ts:239]: If lilconfig is configured with multiple `searchPlaces`, this code will lookup for the same cache entry for each search place. In the case of Prettier, [14 different searchPlaces](https://github.com/prettier/prettier/blob/83b9f62db4111f7d6c3a7ebd2fc01fca237d9072/src/config/get-prettier-config-explorer.js#L11-L25) are passed. 

Since `Map.prototype.has` is very efficient, it shouldn't make a big difference performance-wise. But it be ideal to avoid this.

**Review by pmdartus — COMMENTED** [review]
I know this PR is still in draft. But I thought that you might appreciate some early feedback. 
Very excited about this feature landing. 👏

File: src/index.ts (hunk 1 above)
Comment (antonk52) [reply src/index.ts:240]: Not sure I understand what you mean here.

If you have a config in `./a/foo.config.js` and then call

```js
const searcher = lilconfigSync('foo')

searcher('./a/b/c')
```

it will search towards the root. So the look up loop will run through `./a/b/c`, then `./a/b`, then `./a`, and then `.`.
Once it finds a config it caches all the directories where it looked up configs so now the cache has `./a`, `./a/b`, `./a/b/c`(omitting `.` as we found config before checking it)

When you call it again with a **superset path** like this

```js
searcher('./a/b/c/d')
```

It will search `./a/b/c/d`, cache miss. Then it searches `./a/b/c` -> cache hit, we can return.

At this point I understood your comment. I will update the code. Thanks for pointing it out!

File: src/index.ts (hunk 2 above)
Comment (antonk52) [reply src/index.ts:239]: I am using `searchPath` as a cache key, not `filepath`. On each iteration `searchPath` is the directory where we expect the config to be, not the full path to the config file. This keeps the code simpler and we will have less entries in the cache as a result. Example values from a loop iteration:

`searchPath` -> `/foo/bar/baz`
`filepath` -> `/foo/bar/baz/prettier.config.js`

I agree that the naming is rather confusing. I haven't came up with a better names.

File: src/index.ts (hunk 1 above)
Comment (antonk52) [reply src/index.ts]: fixed in [5ba50ab](https://github.com/antonk52/lilconfig/pull/43/commits/5ba50ab7e1ef99b73a463cc040ae5f3ea6cbf3d5)

File: src/index.ts (hunk 2 above)
Comment (pmdartus) [reply src/index.ts:239]: Thanks for the clarification. I think my comment is still valid, here is a concrete example:

```js
const resolver = lilconfig('config', {
  searchPlaces: ['package.json', '.prettierrc', '.prettierrc.json'],
});

resolver.search('./foo');
```

In the above example, `getSearchItems` would generate the following entries:
* `{ searchPath: './foo', filepath: './foo/package.json' }`
* `{ searchPath: './foo', filepath: './foo/.prettierrc' }`
* `{ searchPath: './foo', filepath: './foo/.prettierrc.json' }`
* `{ searchPath: './', filepath: './package.json' }`
* `{ searchPath: './', filepath: './.prettierrc' }`
* `{ searchPath: './', filepath: './.prettierrc.json' }`

Correct me if I am wrong but with the for loop iterating over the entries generated by `getSearchItems`, when cache is enabled, it appears to me that this code will do:
* 3 consecutive cache lookup with: `searchCache.get('./foo')`
* follow by, 3 consecutive cache lookup with: `searchCache.get('./')`

IMO, it's not a big deal performance-wise since the `has` check on a `Map` is not very expensive. But ideally, the code should do 1 lookup for each `searchpath` in the search cache.

File: src/index.ts [hunk 3]
```diff
@@ -270,28 +324,48 @@ export function lilconfig(
...
+            return emplace(
+                loadCache,
+                absPath,
+                transform(
+                    isEmpty ? {...result, isEmpty, config: undefined} : result,
+                ),
             );
         },
+        clearLoadCache() {
+            if (cache) loadCache.clear();
```
[the remark this answers is not in the document]
Comment (pmdartus) [reply src/index.ts:349]: Sounds good.

File: src/index.ts [hunk 4]
```diff
@@ -101,6 +102,7 @@ function getOptions(
         stopDir: os.homedir(),
         searchPlaces: getDefaultSearchPlaces(name),
         ignoreEmptySearchPlaces: true,
+        cache: true,
```
Comment (pmdartus) [inline src/index.ts:105]: Note: Setting by default `cache` to `true` is certainly a breaking change for downstream consumers. If you want to keep it this way, I would recommend releasing this change under a new major version.

Comment (pmdartus) [reply src/index.ts:105]: >  I consider this change in behaviour a breaking change.

I haven't looked at the PR description. 😂

**Comment by pmdartus** [comment]
> A few more changes will be merged prior to releasing a new version like dropping support for legacy node.js versions.

As a side note, Prettier still supports Node 14 today.

I don't know what is the plan for Prettier to dropping Node 14 and Node 16 (both EOL). I know it's not ideal, but it would be great if lilconfig could still maintain backward compat with Node 14.

File: src/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (antonk52) [reply src/index.ts:313]: I've bench tested the two implementations. you can see the results below. While I do not think that these functions will be called that often. It is nice to see that the performance of it was considered. I will leave the options check as is.

```
cache check x 1,009,410,347 ops/sec ±0.33% (97 runs sampled)
cache clear x 20,670,156 ops/sec ±3.08% (82 runs sampled)
```

**Comment by antonk52** [comment]
I think this PR is ready to be merged. I ran a few benchmarks in `feat/cache-bench` branch. And made a couple of more small touches around getting from and updating the cache. Overall I am happy about the implementation and tests.

There is a small dip in test coverage because of an edge case that only occurs in nixos. I am not sure how this can be tested, so ignoring it for now.


While I am confident about basic benchmarks with set and map, the benchmarks with searches seem ludicrous to me. I might have messed up something with benchmark.js. I have no doubt about the new implementation being faster than the old one, but these numbers show too much of a difference to be believable.

```
Running "sync single search"
old x 959 ops/sec ±1.01% (93 runs sampled)
new x 31,080,130 ops/sec ±0.36% (97 runs sampled)
Fastest is new
```

File: src/index.ts (hunk 2 above)
Comment (antonk52) [reply src/index.ts]: Thanks for expanding on this. I've switched the look up for two nested loops. No more unnecessary look ups
